### PR TITLE
[python] Refactor output formatter for rolling batch

### PR DESCRIPTION
**Pull request by xyang16**
## Description ##

Brief description of what this PR is about

- If this change is a backward incompatible change, why must this change be made?
- Interesting edge cases to note here

### CODE REVIEW

File: engines/python/setup/djl_python/huggingface.py [hunk 1]
```diff
@@ -173,8 +173,6 @@ def initialize(self, properties: dict):
                 self.device = int(os.getenv("LOCAL_RANK", 0))
             _rolling_batch_cls = get_rolling_batch_class_from_str(
                 self.rolling_batch_type, is_mpi, self.model_config)
-
-            # TODO: Allow user to set output formatter
             self.rolling_batch = _rolling_batch_cls(model_id_or_path,
```
Comment (lanking520) [inline engines/python/setup/djl_python/huggingface.py:176]: you can make output formatter as part of the kwargs, user can pass in a function if needed

File: engines/python/setup/djl_python/rolling_batch/rolling_batch.py [hunk 2]
```diff
@@ -124,6 +109,13 @@ def __init__(self, device):
         self.device = device
         self.pending_requests = []
         self.req_id_counter = 0
+        self.content_type = None
+        self.output_formatter = None
+
+    def init(self, content_type="application/jsonlines"):
```
Comment (lanking520) [inline engines/python/setup/djl_python/rolling_batch/rolling_batch.py:115]: why we need a separate init function

File: engines/python/setup/djl_python/huggingface.py (hunk 1 above)
Comment (xyang16) [reply engines/python/setup/djl_python/huggingface.py:176]: Changed.

File: engines/python/setup/djl_python/rolling_batch/rolling_batch.py (hunk 2 above)
Comment (xyang16) [reply engines/python/setup/djl_python/rolling_batch/rolling_batch.py]: Changed.

File: engines/python/setup/djl_python/rolling_batch/scheduler_rolling_batch.py [hunk 3]
```diff
@@ -143,7 +143,13 @@ def _prefill_and_decode(self, new_requests):
         for request_id, generated_token, request in zip(
                 request_ids, generated_tokens, self.pending_requests):
             is_last_token = (request_id in exit_req_ids)
-            request.set_next_token(generated_token, last_token=is_last_token)
+            if self.output_formatter is not None:
```
Comment (frankfliu) [inline engines/python/setup/djl_python/rolling_batch/scheduler_rolling_batch.py:146]: request.set_next_token(generated_token, self.output_formatter, last_token=is_last_token)

Comment (xyang16) [reply engines/python/setup/djl_python/rolling_batch/scheduler_rolling_batch.py:146]: Do not format output for scheduler_rolling_batch?

File: engines/python/setup/djl_python/rolling_batch/rolling_batch.py [hunk 4]
```diff
@@ -114,16 +99,22 @@ class RollingBatch(ABC):
...
         :param device: device to load the model
+        :param kwargs passed while loading the model
         """
 
         self.device = device
         self.pending_requests = []
         self.req_id_counter = 0
+        if 'rolling_batch_output_formatter' in kwargs:
+            # TODO: Allow user to set custom output formatter
+            pass
```
Comment (lanking520) [inline engines/python/setup/djl_python/rolling_batch/rolling_batch.py:115]: self.output_formatter = kwargs['rolling_batch_output_formatter']

File: engines/python/setup/djl_python/rolling_batch/scheduler_rolling_batch.py (hunk 3 above)
Comment (xyang16) [reply engines/python/setup/djl_python/rolling_batch/scheduler_rolling_batch.py]: Changed

File: engines/python/setup/djl_python/rolling_batch/rolling_batch.py [hunk 5]
```diff
@@ -169,3 +160,9 @@ def postprocess_results(self, batch_size):
             self.req_id_counter = 0
 
         return results
+
+    def get_content_type(self):
+        # TODO: find a way to return content-type for custom output formatter
+        if self.output_formatter == _default_output_formatter:
+            return "application/jsonlines"
+        return None
```
Comment (lanking520) [inline engines/python/setup/djl_python/rolling_batch/rolling_batch.py:174]: can be plain text?

File: engines/python/setup/djl_python/rolling_batch/rolling_batch.py (hunk 4 above)
Comment (xyang16) [reply engines/python/setup/djl_python/rolling_batch/rolling_batch.py:121]: Changed

**Review by lanking520 — APPROVED** [review]

File: engines/python/setup/djl_python/rolling_batch/rolling_batch.py (hunk 5 above)
Comment (xyang16) [reply engines/python/setup/djl_python/rolling_batch/rolling_batch.py:174]: Maybe some other types

**Comment by xyang16** [comment]
Reopen

**Comment by xyang16** [comment]
Reopen

**Pull request merged**